sdk/core/azure-core/src/main/java/com/azure/core/util/ConfigurationProperty.java

```diff
@@ -0,0 +1,118 @@
+
+
+/**
+ * Represents configuration property.
+ *
+ * @param <T> Type of property value.
+ */
+public class ConfigurationProperty<T> {
+    private static final String[] EMPTY_ARRAY = new String[0];
+    private static final Function<String, String> REDACT_VALUE_SANITIZER = (value) -> "redacted";
+
+    private final String name;
+    private final String[] aliases;
+    private final String environmentVariable;
+    private final String systemProperty;
+    private final Function<String, T> converter;
+    private final Function<String, String> valueSanitizer;
+    private final T defaultValue;
+    private final boolean isShared;
+    private final boolean isRequired;
+
+    /**
+     * Creates configuration property.
+     *
+     * @param name full name (including relative path) of the property.
+     * @param defaultValue default value.
+     * @param isRequired indicates required property for validation purposes
+     * @param converter converts String property value to property type and performs validation.
+     * @param isShared indicates property that can be specified in per-client and root sections.
+     * @param environmentVariable environment variables the property can be represented with. Can be null.
+     * @param systemProperty system property the property can be represented with. Can be null.
+     * @param aliases list of alternative names of the property. Can be null.
+     * @param valueSanitizer sanitizes property value for logging purposes.
+     */
+    ConfigurationProperty(String name, T defaultValue, boolean isRequired, Function<String, T> converter, boolean isShared,
+                          String environmentVariable, String systemProperty, String[] aliases, Function<String, String> valueSanitizer) {
+        this.name = Objects.requireNonNull(name, "'name' cannot be null");
+        this.converter = Objects.requireNonNull(converter, "'converter' cannot be null");
+        this.environmentVariable = environmentVariable;
+        this.systemProperty = systemProperty;
+        this.aliases = aliases == null ? EMPTY_ARRAY : aliases;
+        this.defaultValue = defaultValue;
+        this.isRequired = isRequired;
+        this.isShared = isShared;
+        this.valueSanitizer = valueSanitizer == null ? REDACT_VALUE_SANITIZER : valueSanitizer;
+    }
+
+    /**
+     * Returns true if property can be shared between clients and configuration should
+     * look for it in per-client and root sections.
+     */
+    boolean isShared() {
+        return isShared;
+    }
+
+    /**
+     * Returns true if property value can be securely logged.
+     */
+    Function<String, String> getValueSanitizer() {
+        return valueSanitizer;
+    }
+
+    /**
+     * Returns true if property is required, used for validation purposes.
+     */
+    boolean isRequired() {
+        return isRequired;
+    }
+
+    /**
+     * Gets full property name including relative path to it.
+     */
+    String getName() {
```

Comment:
Since these are read-only methods, we may want to consider making these public for anyone to inspect name, `isShared`, `isRequired`, `canLogValue` settings for a property.